src/core/mac/mac_frame.cpp

```diff
@@ -734,34 +768,71 @@ void Frame::SetPayloadLength(uint16_t aLength)
 
 uint8_t Frame::SkipSecurityHeaderIndex(void) const
 {
-    uint8_t  index = 0;
+    uint8_t  index = kFcfSize + kDsnSize; // Frame Control field and Sequence Number field
     uint16_t fcf;
 
-    // Frame Control
-    index += kFcfSize;
-    // Sequence Number
-    index += kDsnSize;
-
     VerifyOrExit((index + GetFcsSize()) <= GetPsduLength(), index = kInvalidIndex);
 
     fcf = GetFrameControlField();
+    SkipAddrField(fcf, index);
+
+    // Security Control + Frame Counter + Key Identifier
+    if ((fcf & kFcfSecurityEnabled) != 0)
+    {
+        VerifyOrExit(index < GetPsduLength(), index = kInvalidIndex);
+        uint8_t securityControl = *(GetPsdu() + index);
+
+        index += kSecurityControlSize + kFrameCounterSize;
+
+        switch (securityControl & kKeyIdModeMask)
+        {
+        case kKeyIdMode0:
+            index += kKeySourceSizeMode0;
+            break;
+
+        case kKeyIdMode1:
+            index += kKeySourceSizeMode1 + kKeyIndexSize;
+            break;
 
-    // Destination PAN + Address
+        case kKeyIdMode2:
+            index += kKeySourceSizeMode2 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode3:
+            index += kKeySourceSizeMode3 + kKeyIndexSize;
+            break;
+        }
+    }
+
+exit:
+    return index;
+}
+
+void Frame::SkipAddrField(uint16_t fcf, uint8_t &index) const
```

Comment:
Can we consider adopting a model like `SkipSecurityHeaderIndex()` returning `uint8_t` from this function.

We pass in an `index` to this method  but if I am correct, in all cases, the index is set to `index = kFcfSize + kDsnSize` before calling this. Is there any situation where this may be different?